File: day2/node/main.js
const fs = require('fs')
const assert = require('assert');

const validate_password_method_one = (password, pattern) => {
   const pattern_parts = pattern.split(' ');
   const character = pattern_parts[1];
   const x = pattern_parts[0].split('-')[0];
   const y = pattern_parts[0].split('-')[1];

   const regex = new RegExp(character,"g")
   const occurrences = password.match(regex) ? password.match(regex).length : 0;
   return occurrences >= x && occurrences <= y;
}

const validate_password_method_two = (password, pattern) => {
    const pattern_parts = pattern.split(' ');
    const character = pattern_parts[1];
    const x = pattern_parts[0].split('-')[0];
    const y = pattern_parts[0].split('-')[1];
    const character_in_x = password.charAt(x - 1);
    const character_in_y = password.charAt(y - 1);

    if (character_in_x === character_in_y && character_in_x === character) {
        return false;
    }

    if (character_in_x === character || character_in_y === character) {
        return true
    }

    return false;
}

const main = (filepath, validation_method) => {
    return new Promise((resolve, reject) => {
        fs.readFile(filepath, 'utf8', function (err, data) {
            if (err) {
                reject(err);
            }

            const patterns_and_passwords_list = data.toString().split("\n");
            let passwords_ok = 0

            patterns_and_passwords_list.forEach((pattern_and_password) => {
                const pattern_and_password_parts = pattern_and_password.split(':')
                if (validation_method(pattern_and_password_parts[1].trimStart(), pattern_and_password_parts[0].trimStart())) {
                    passwords_ok++;
                }
            })

            resolve(passwords_ok);
        });
    });
}

const start1 = process.hrtime();
main('../inputs/test.txt', validate_password_method_one).then(result => {
    const hrend = process.hrtime(start1)

    assert.strictEqual(result, 2);
    console.log('- Test (method 1): ' + result + ' [%dms]', hrend[1] / 1000000000);
});

const start2 = process.hrtime();
main('../inputs/test.txt', validate_password_method_two).then(result => {
    const hrend = process.hrtime(start2)

    assert.strictEqual(result, 1);
    console.log('- Test (method 2): ' + result + ' [%dms]', hrend[1] / 1000000000);
});

const start3 = process.hrtime();
main('../inputs/input.txt', validate_password_method_one).then(result => {
    const hrend = process.hrtime(start3)
    console.log('- Part One (method 1): ' + result + ' [%dms]', hrend[1] / 1000000000);
});

const start4 = process.hrtime();
main('../inputs/input.txt', validate_password_method_two).then(result => {
    const hrend = process.hrtime(start3)
    console.log('- Part Two (method 2): ' + result + ' [%dms]', hrend[1] / 1000000000);
});

File: day2/python/main.py
import time


def readfile(filepath):
    """
    This function reads the file associated with filename and returns a list of patterns

    :param filepath: The path of the file containing the list of numbers
    :return: returns list of patterns
    """
    fileObj = open(filepath, "r")
    patterns = fileObj.read().splitlines()
    fileObj.close()
    return patterns


def validate_password_method_one(password, pattern):
    """
    This function validates the param password against the pattern:
    The lowest and highest number of times a given letter must appear for the password

    :param password: password to validate
    :param pattern: pattern used to validate the password
    :return: true/false whether the password is right or wrong
    """
    pattern_parts = pattern.split()
    character = pattern_parts[1]
    x = int(pattern_parts[0].split('-')[0])
    y = int(pattern_parts[0].split('-')[1])

    occurrences = password.count(character)

    return x <= occurrences <= y


def validate_password_method_two(password, pattern):
    """
    This function validates the param password against the pattern:
    Exactly one of positions given in the pattern must contain the given letter

    :param password: password to validate
    :param pattern: pattern used to validate the password
    :return: true/false whether the password is right or wrong
    """
    pattern_parts = pattern.split()

    x = int(pattern_parts[0].split('-')[0])
    y = int(pattern_parts[0].split('-')[1])
    character_expected = pattern_parts[1]
    try:
        character_in_x = password[x - 1]
        character_in_y = password[y - 1]
        if character_in_x == character_expected == character_in_y:
            return False

        if character_in_x == character_expected or character_in_y == character_expected:
            return True

    except IndexError:
        return False

    return False


def main(file, validation_method):
    """
    Main method
    :param file: file containing patterns andd passwords
    :param validation_method: method used to validate passwords
    :return:
    """
    patterns_and_passwords_list = readfile(file)
    passwords_ok = 0
    for pattern_and_password in patterns_and_passwords_list:
        pattern_and_password_parts = pattern_and_password.split(':')
        if validation_method(pattern_and_password_parts[1].strip(), pattern_and_password_parts[0].strip()):
            passwords_ok += 1

    return passwords_ok


## Asserts
assert main('../inputs/test.txt', validate_password_method_one) == 2
assert main('../inputs/test.txt', validate_password_method_two) == 1

start_time = time.time()
print(
    "- Part One (method 1): " + str(main('../inputs/input.txt', validate_password_method_one)) + " ok [%s seconds]" % (
                time.time() - start_time))

start_time = time.time()
print(
    "- Part Two (method 2): " + str(main('../inputs/input.txt', validate_password_method_two)) + " ok [%s seconds]" % (
                time.time() - start_time))
